### PR TITLE
Use real HTTP bind address in curl RPC help

### PR DESCRIPTION
**Pull request by domob1812**
This change exposes the bind address (one of them if multiple) of the HTTP server.  This allows us to use a real bind address in the `curl` example in RPC help texts.

In particular, this now reports the correct port if it was either explicitly changed through `-rpcport` or is from a different chain than mainnet.

This is a replacement for #15181, based on the feedback from @laanwj there.

### CODE REVIEW

**Comment by domob1812** [comment]
This includes some trivial refactoring (`static` replaced by anonymous namespaces) in places that I changed anyway - I think this would not be worth doing in a separate PR.  But of course I'm happy to either remove those changes or submit them in a separate commit / PR if you think that's better.

File: src/httpserver.cpp [hunk 1]
```diff
@@ -478,6 +490,15 @@ void StopHTTPServer()
     LogPrint(BCLog::HTTP, "Stopped HTTP server\n");
 }
 
+bool GetHTTPBindAddress(std::string& address)
+{
+    if (bindAddress.empty()) {
+        return false;
```
Comment (promag) [inline src/httpserver.cpp]:
```cpp
address = bindAddress;
return !address.empty();
```

**Review by promag — COMMENTED** [review]
This has unnecessary/unrelated changes.

File: src/httpserver.cpp (hunk 1 above)
Comment (domob1812) [reply src/httpserver.cpp]: Yes, but that would overwrite the passed in address.  While it doesn't matter in the current context, I found it cleaner to not affect it at all if `false` is returned.

**Comment by domob1812** [comment]
> This has unnecessary/unrelated changes.

See my comment above.  I feel those changes are an improvement and too insignificant to do on their own, but I'm happy to either split them out or not do them at all.

**Comment by promag** [comment]
At least have in a separate commit. If they are an improvement and unrelated here then a separate PR makes more sense.

**Comment by domob1812** [comment]
> At least have in a separate commit. If they are an improvement and unrelated here then a separate PR makes more sense.

Agreed, I've split it into two commits for now.  As I wrote above, I'm happy to drop the first one or submit in a separate PR if you think that makes more sense.

**Comment by MarcoFalke** [comment]
This is a lot of code for making an example more "real". Generally, I'd prefer if the example was static, since it ends up on static websites (https://bitcoincore.org/en/doc/0.16.0/rpc/rawtransactions/getrawtransaction/ or http://chainquery.com/bitcoin-api/getrawtransaction)

Wouldn't it make more sense to include that information in the getrpcinfo command as suggested by @instagibbs https://github.com/bitcoin/bitcoin/pull/14982#issuecomment-449412662?

**Comment by domob1812** [comment]
> This is a lot of code for making an example more "real". Generally, I'd prefer if the example was static, since it ends up on static websites (https://bitcoincore.org/en/doc/0.16.0/rpc/rawtransactions/getrawtransaction/ or http://chainquery.com/bitcoin-api/getrawtransaction)
> 
> Wouldn't it make more sense to include that information in the getrpcinfo command as suggested by @instagibbs [#14982 (comment)](https://github.com/bitcoin/bitcoin/pull/14982#issuecomment-449412662)?

I don't know.  I just noticed that the port can be plain wrong in the static example, for instance if running in regtest mode.  (Which is actually something that developers might do and where they might want to use the `curl` example to access their node!)

@laanwj's last opinion was that he would be happy with that change in the form it is now, at least if I understood him correctly.  But since he would also be happy not doing the fix and I don't care too strongly about it (especially with the amount of back and forth this is getting to), I'm happy to just drop this attempt finally.

**Comment by promag** [comment]
Agree with @MarcoFalke, the help text is an *example*.

File: test/functional/rpc_help.py [hunk 2]
```diff
@@ -43,6 +45,10 @@ def test_categories(self):
 
         assert_equal(titles, components)
 
+    def test_exampleBindAddress(self):
```
Comment (practicalswift) [inline test/functional/rpc_help.py]: Should be `snake_case`? :-)

File: test/functional/rpc_help.py [hunk 3]
```diff
@@ -43,6 +45,10 @@ def test_categories(self):
...
+        helpText = self.nodes[0].help('getblockchaininfo')
```
Comment (practicalswift) [inline test/functional/rpc_help.py]: Same here -- should be `snake_case`? :-)

**Comment by laanwj** [comment]
> I've fixed the two nits found by @practicalswift.  But if the general opinion is that we should just not do this, then feel free to just close the PR.

Tend to agree. This just keeps dragging on.

But it's better to do that yourself! (than you still have control over it)

**Comment by domob1812** [comment]
> > I've fixed the two nits found by @practicalswift.  But if the general opinion is that we should just not do this, then feel free to just close the PR.
> 
> Tend to agree. This just keeps dragging on.
> 
> But it's better to do that yourself! (than you still have control over it)

Sounds good, then just let me close it. :)

**Pull request closed without merging**